src/bp/ml/crf-js/index.ts

```diff
@@ -0,0 +1,77 @@
+import * as sdk from 'botpress/sdk'
+import nanoid from 'nanoid'
+
+import addon, { Tagger as AddonTagger, Trainer as AddonTrainer, Options } from './addon'
+
+export class Trainer implements Omit<AddonTrainer, 'train'>, sdk.MLToolkit.CRF.Trainer {
+  private crfsuite_addon: AddonTrainer
+
+  constructor() {
+    this.crfsuite_addon = new addon.Trainer({ debug: true })
+  }
+
+  append(xseq: string[][], yseq: string[]): void {
+    return this.crfsuite_addon.append(xseq, yseq)
+  }
+
+  train_async(model_filename: string, cb: (s: string) => void): Promise<number> {
```

Comment:
I don't understand how that works ? it trains from file ? or is it that it trains from all the records you've set with .append and it outputs to model_filename ? 

What is the number that it returns ?

Reply:
The c++ code of the sync version of this function looks like this:
```
Napi::Value TrainerClass::Train(const Napi::CallbackInfo &info)
{
  if (info.Length() < 1 || !info[0].IsString())
  {
    Napi::TypeError::New(info.Env(), "Path to model file is missing or invalid").ThrowAsJavaScriptException();
  }

  Napi::String path = info[0].As<Napi::String>();

  int32_t status = this->trainer->train(path.Utf8Value(), -1);
  return Napi::Number::New(info.Env(), status);
}
```
So it takes as parameter a model filename and outputs a training status.

The callback is only called with debbuging message from crfsuite itself.

I honnestly don't know anything more, I have'nt digged deeper in crfsuite...